examples/fx_gan.py

```diff
@@ -0,0 +1,144 @@
+class Discriminator(nn.Module):
+    def __init__(self, channels = 1, out_channels = 1, gap_size = feat_length, noise = 0.1):
+        super(Discriminator, self).__init__()
+        self.noise = noise
+        self.conv1 = base_conv1d(in_channels = channels, out_channels = 512, ks = 7, pd = 3)
+        self.conv2 = base_conv1d(in_channels = 512, out_channels = 256, ks = 5, pd = 2)
+        self.conv3 = base_conv1d(in_channels = 256, out_channels = 128)
+        self.out = out_layer(in_channels = 128, out_channels = out_channels, gap_size = gap_size)
+    def forward(self, x):
+        if self.training and self.noise != 0.0:
+            x = x + Variable(x.data.new(x.size()).normal_(0.0, self.noise))
+        return self.out(self.conv3(self.conv2(self.conv1(x))))
+
+# Generator needs to take the latent variables and then generate a fake sequence of given channels
+class Generator(nn.Module):
+    def __init__(self, channels = 1, gap_size = feat_length):
+        super(Generator, self).__init__()
+        self.channels = channels
+        self.gap_size = gap_size
+        self.conv1 = base_conv1d(in_channels = 128, out_channels = 256)
+        self.conv2 = base_conv1d(in_channels = 256, out_channels = 512, ks = 5, pd = 2)
+        self.conv3 = base_conv1d(in_channels = 512, out_channels = feat_length-input_size, ks = 7, pd = 3)
+        self.conv3.activation = nn.Sigmoid() # Because the input is in [0, 1]
+    def forward(self, x, direction = True):
+        gen = Variable(torch.randn(x.shape[0], 128, self.channels)).cuda()
+        gen = self.conv3(self.conv2(self.conv1(gen))).view(-1, self.channels, feat_length-input_size)
+        return torch.cat([gen, x], dim=2) if direction else torch.cat([x, gen], dim=2)
+
+Dataset = pd.read_csv('Dataset.csv', index_col='Gmt time', parse_dates=True, infer_datetime_format=True)
+Dataset = Dataset.drop(['Imputated'], axis=1)
+
+input_vectors = []
+
+for i in range(len(Dataset)//feat_length):
+    start = i * feat_length
+    end = start + input_size
+    input_vector = Dataset.iloc[start:end].values
+    input_vector = np.swapaxes(input_vector, 0, 1)
+    input_vectors.append(input_vector)
+
+X = np.asarray(input_vectors)
+
+X_transformer = QuantileTransformer(output_distribution='uniform')
+X = X_transformer.fit_transform(X.reshape(-1, input_size))
+X_train = torch.from_numpy(X.reshape(-1, feat_channel, input_size))
+
+train_dataset = torch.utils.data.TensorDataset(X_train)
+
+train_loader = torch.utils.data.DataLoader(train_dataset, 
+    batch_size = 64, shuffle = True, pin_memory = True)
+
+G = Generator(channels = feat_channel, gap_size = feat_length).cuda()
+D = Discriminator(channels = feat_channel, gap_size = feat_length).cuda()
+
+G_opt = torch.optim.Adam(params = G.parameters(), lr = 1e-3)
+D_opt = torch.optim.Adam(params = D.parameters(), lr = 1e-3)
+
+BCE = torch.nn.BCELoss()
+
+G.train(True)
+D.train(True)
+
+for epoch in range(epochs):
```

Comment:
This train loop is pretty trivial and can be handled directly using the `Trainer`. You just need to define the models and losses as I have described then feed them to the Trainer. You don't need to implement any plotting explicitly those are already taken care of.